File: tests/runners/cpp/tests/test_08_complex_schema.cpp
// Test Case 08: Complex Schema
// Comprehensive test combining all features

#include <iostream>
#include <cassert>
#include <cmath>

#include "schema.hpp"
#include "schema_loaders.hpp"
#include "schema_container.hpp"

void test_common_types() {
    std::cout << "  Testing common types (Vec2, Vec3, Color, Element)..." << std::endl;

    game::common::Vec2 v2;
    v2.x = 1.0f;
    v2.y = 2.0f;
    assert(v2.x == 1.0f);
    assert(v2.y == 2.0f);

    game::common::Vec3 v3;
    v3.x = 1.0f;
    v3.y = 2.0f;
    v3.z = 3.0f;
    assert(v3.z == 3.0f);

    game::common::Color color;
    color.r = 255;
    color.g = 128;
    color.b = 64;
    color.a = 255;
    assert(color.r == 255);
    assert(color.g == 128);

    game::common::Element elem = game::common::Element::Fire;
    assert(static_cast<int32_t>(elem) == 1);

    std::cout << "    PASS" << std::endl;
}

void test_character_types() {
    std::cout << "  Testing character types (Stats, Player, NPC)..." << std::endl;

    // Stats
    game::character::Stats stats;
    stats.hp = 100;
    stats.max_hp = 100;
    stats.mp = 50;
    stats.max_mp = 50;
    stats.strength = 10;
    stats.agility = 8;
    stats.intelligence = 5;
    stats.vitality = 12;

    // Player
    game::character::Player player;
    player.id = 1;
    player.name = "Hero";
    player.level = 10;
    player.experience = 5000;
    player.stats = stats;
    player.position.x = 100.0f;
    player.position.y = 50.0f;
    player.position.z = 0.0f;
    player.status = game::character::Player::Status::Online;
    player.guild_id = 1;

    assert(player.name == "Hero");
    assert(player.level == 10);
    assert(player.stats.strength == 10);
    assert(player.guild_id.has_value());
    assert(*player.guild_id == 1);

    // NPC
    game::character::NPC npc;
    npc.id = 1;
    npc.name = "Merchant";
    npc.title = "Item Seller";
    npc.stats = stats;
    npc.spawn_position.x = 50.0f;
    npc.spawn_position.y = 50.0f;
    npc.spawn_position.z = 0.0f;
    npc.ai_type = game::character::NPC::AIType::Friendly;

    game::character::DialogOption opt1, opt2;
    opt1.text = "Hello!";
    opt1.next_dialog_id = 2;
    opt2.text = "Goodbye!";
    opt2.next_dialog_id = std::nullopt;
    npc.dialog_options = {opt1, opt2};

    assert(npc.title.has_value());
    assert(*npc.title == "Item Seller");
    assert(npc.dialog_options.size() == 2);
    assert(npc.dialog_options[0].next_dialog_id.has_value());
    assert(!npc.dialog_options[1].next_dialog_id.has_value());

    std::cout << "    PASS" << std::endl;
}

void test_item_types() {
    std::cout << "  Testing item types (Item, Weapon, Armor)..." << std::endl;

    // Item
    game::item::Item item;
    item.id = 1;
    item.name = "Iron Sword";
    item.description = "A basic sword";
    item.rarity = game::item::Rarity::Common;
    item.sell_price = 100;
    item.max_stack = 1;
    item.icon = "sword_01";
    item.item_type = game::item::Item::ItemType::Weapon;

    assert(item.name == "Iron Sword");
    assert(item.rarity == game::item::Rarity::Common);

    // Weapon
    game::item::Weapon weapon;
    weapon.item_id = 1;
    weapon.damage_min = 10;
    weapon.damage_max = 15;
    weapon.attack_speed = 1.2f;
    weapon.element = game::common::Element::Physical;
    weapon.equip_slot = game::character::EquipSlot::MainHand;

    game::item::BonusStat bonus;
    bonus.stat_name = "Strength";
    bonus.value = 5;
    weapon.bonus_stats = {bonus};

    assert(weapon.damage_min == 10);
    assert(weapon.bonus_stats.size() == 1);
    assert(weapon.bonus_stats[0].value == 5);

    // Armor
    game::item::Armor armor;
    armor.item_id = 2;
    armor.defense = 20;
    armor.magic_defense = 10;
    armor.equip_slot = game::character::EquipSlot::Body;

    game::item::Resistance res1, res2;
    res1.element = game::common::Element::Fire;
    res1.value = 10;
    res2.element = game::common::Element::Ice;
    res2.value = -5;  // weakness
    armor.resistances = {res1, res2};

    assert(armor.defense == 20);
    assert(armor.resistances.size() == 2);
    assert(armor.resistances[1].value == -5);

    std::cout << "    PASS" << std::endl;
}

void test_inventory_system() {
    std::cout << "  Testing inventory system..." << std::endl;

    game::inventory::InventorySlot slot;
    slot.id = 1;
    slot.player_id = 1;
    slot.item_id = 1;
    slot.slot_index = 0;
    slot.quantity = 5;

    game::inventory::Enhancement enh;
    enh.level = 3;
    enh.bonus_value = 15;
    slot.enhancement = enh;

    assert(slot.quantity == 5);
    assert(slot.enhancement.has_value());
    assert(slot.enhancement->level == 3);

    game::inventory::Equipment equip;
    equip.player_id = 1;
    equip.slot = game::character::EquipSlot::MainHand;
    equip.item_id = 1;

    assert(equip.slot == game::character::EquipSlot::MainHand);

    std::cout << "    PASS" << std::endl;
}

void test_social_system() {
    std::cout << "  Testing social system (Guild, GuildMember, Friendship)..." << std::endl;

    game::social::Guild guild;
    guild.id = 1;
    guild.name = "Heroes";
    guild.tag = "HRO";
    guild.leader_id = 1;
    guild.level = 5;
    guild.emblem_color.r = 255;
    guild.emblem_color.g = 215;
    guild.emblem_color.b = 0;
    guild.emblem_color.a = 255;
    guild.created_at = 1640000000;

    assert(guild.name == "Heroes");
    assert(guild.tag == "HRO");
    assert(guild.emblem_color.r == 255);

    game::social::GuildMember member;
    member.guild_id = 1;
    member.player_id = 1;
    member.rank = game::social::GuildMember::Rank::Leader;
    member.joined_at = 1640000000;

    assert(member.rank == game::social::GuildMember::Rank::Leader);

    game::social::Friendship friendship;
    friendship.player_a_id = 1;
    friendship.player_b_id = 2;
    friendship.since = 1641000000;

    assert(friendship.player_a_id == 1);

    std::cout << "    PASS" << std::endl;
}

void test_container_integration() {
    std::cout << "  Testing container integration..." << std::endl;

    schema_container::SchemaContainer container;

    // Add player
    game::character::Player player;
    player.id = 1;
    player.name = "TestPlayer";
    player.level = 50;
    player.experience = 100000;
    player.stats.hp = 500;
    player.stats.max_hp = 500;
    player.stats.mp = 200;
    player.stats.max_mp = 200;
    player.stats.strength = 50;
    player.stats.agility = 40;
    player.stats.intelligence = 30;
    player.stats.vitality = 60;
    player.position.x = 0;
    player.position.y = 0;
    player.position.z = 0;
    player.status = game::character::Player::Status::Online;
    player.guild_id = std::nullopt;

    container.players.add_row(player);

    // Add item
    game::item::Item item;
    item.id = 1;
    item.name = "Legendary Sword";
    item.description = "A sword of legends";
    item.rarity = game::item::Rarity::Legendary;
    item.sell_price = 10000;
    item.max_stack = 1;
    item.icon = "legendary_sword";
    item.item_type = game::item::Item::ItemType::Weapon;

    container.items.add_row(item);

    // Verify lookups
    auto* found_player = container.players.get_by_name("TestPlayer");
    assert(found_player != nullptr);
    assert(found_player->level == 50);

    auto* found_item = container.items.get_by_name("Legendary Sword");
    assert(found_item != nullptr);
    assert(found_item->rarity == game::item::Rarity::Legendary);

    std::cout << "    PASS" << std::endl;
}

void test_binary_complex() {
    std::cout << "  Testing binary serialization of complex types..." << std::endl;

    game::character::Player original;
    original.id = 999;
    original.name = "SerializationTest";
    original.level = 99;
    original.experience = 9999999;
    original.stats.hp = 9999;
    original.stats.max_hp = 9999;
    original.stats.mp = 4999;
    original.stats.max_mp = 4999;
    original.stats.strength = 255;
    original.stats.agility = 255;
    original.stats.intelligence = 255;
    original.stats.vitality = 255;
    original.position.x = 123.456f;
    original.position.y = 789.012f;
    original.position.z = 345.678f;
    original.status = game::character::Player::Status::InBattle;
    original.guild_id = 42;

    // Serialize
    std::vector<uint8_t> buffer;
    {
        polygen::BinaryWriter writer(buffer);
        polygen_loaders::write_Player(writer, original);
    }

    // Deserialize
    game::character::Player loaded;
    {
        polygen::BinaryReader reader(buffer);
        loaded = polygen_loaders::read_Player(reader);
    }

    assert(loaded.id == original.id);
    assert(loaded.name == original.name);
    assert(loaded.level == original.level);
    assert(loaded.experience == original.experience);
    assert(loaded.stats.hp == original.stats.hp);
    assert(loaded.stats.strength == original.stats.strength);
    assert(std::abs(loaded.position.x - original.position.x) < 0.001f);
    assert(loaded.status == original.status);
    assert(loaded.guild_id.has_value());
    assert(*loaded.guild_id == 42);

    std::cout << "    PASS (serialized " << buffer.size() << " bytes)" << std::endl;
}

int main() {
    std::cout << "=== Test Case 08: Complex Schema ===" << std::endl;

    test_common_types();
    test_character_types();
    test_item_types();
    test_inventory_system();
    test_social_system();
    test_container_integration();
    test_binary_complex();

    std::cout << "=== All tests passed! ===" << std::endl;
    return 0;
}

File: tests/runners/csharp/tests/Test_08_complex_schema.cs
// Test Case 08: Complex Schema
// Comprehensive test combining all features

using System;
using System.Collections.Generic;

class Program
{
    static int passed = 0;
    static int failed = 0;

    static void Assert(bool condition, string message)
    {
        if (!condition)
        {
            Console.WriteLine($"    FAILED: {message}");
            failed++;
        }
    }

    static void TestCommonTypes()
    {
        Console.WriteLine("  Testing common types (Vec2, Vec3, Color, Element)...");

        var v2 = new game.common.Vec2 { x = 1.0f, y = 2.0f };
        Assert(v2.x == 1.0f, "v2.x");
        Assert(v2.y == 2.0f, "v2.y");

        var v3 = new game.common.Vec3 { x = 1.0f, y = 2.0f, z = 3.0f };
        Assert(v3.z == 3.0f, "v3.z");

        var color = new game.common.Color { r = 255, g = 128, b = 64, a = 255 };
        Assert(color.r == 255, "color.r");
        Assert(color.g == 128, "color.g");

        var elem = game.common.Element.Fire;
        Assert((int)elem == 1, "Element.Fire");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestCharacterTypes()
    {
        Console.WriteLine("  Testing character types (Stats, Player, NPC)...");

        var stats = new game.character.Stats
        {
            hp = 100, max_hp = 100,
            mp = 50, max_mp = 50,
            strength = 10, agility = 8, intelligence = 5, vitality = 12
        };

        var player = new game.character.Player
        {
            id = 1,
            name = "Hero",
            level = 10,
            experience = 5000,
            stats = stats,
            position = new game.common.Vec3 { x = 100.0f, y = 50.0f, z = 0.0f },
            status = game.character.Player.Status.Online,
            guild_id = 1
        };

        Assert(player.name == "Hero", "player.name");
        Assert(player.level == 10, "player.level");
        Assert(player.stats.strength == 10, "player.stats.strength");
        Assert(player.guild_id == 1, "player.guild_id");

        var npc = new game.character.NPC
        {
            id = 1,
            name = "Merchant",
            title = "Item Seller",
            stats = stats,
            spawn_position = new game.common.Vec3 { x = 50.0f, y = 50.0f, z = 0.0f },
            ai_type = game.character.NPC.AIType.Friendly,
            dialog_options = new List<game.character.NPC.DialogOption>
            {
                new game.character.NPC.DialogOption { text = "Hello!", next_dialog_id = 2 },
                new game.character.NPC.DialogOption { text = "Goodbye!", next_dialog_id = null }
            }
        };

        Assert(npc.title == "Item Seller", "npc.title");
        Assert(npc.dialog_options.Count == 2, "npc.dialog_options.Count");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestItemTypes()
    {
        Console.WriteLine("  Testing item types (Item, Weapon, Armor)...");

        var item = new game.item.Item
        {
            id = 1,
            name = "Iron Sword",
            description = "A basic sword",
            rarity = game.item.Rarity.Common,
            sell_price = 100,
            max_stack = 1,
            icon = "sword_01",
            item_type = game.item.Item.ItemType.Weapon
        };

        Assert(item.name == "Iron Sword", "item.name");
        Assert(item.rarity == game.item.Rarity.Common, "item.rarity");

        var weapon = new game.item.Weapon
        {
            item_id = 1,
            damage_min = 10,
            damage_max = 15,
            attack_speed = 1.2f,
            element = game.common.Element.Physical,
            equip_slot = game.character.EquipSlot.MainHand,
            bonus_stats = new List<game.item.Weapon.BonusStat>
            {
                new game.item.Weapon.BonusStat { stat_name = "Strength", value = 5 }
            }
        };

        Assert(weapon.damage_min == 10, "weapon.damage_min");
        Assert(weapon.bonus_stats.Count == 1, "weapon.bonus_stats.Count");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestSocialSystem()
    {
        Console.WriteLine("  Testing social system (Guild, GuildMember, Friendship)...");

        var guild = new game.social.Guild
        {
            id = 1,
            name = "Heroes",
            tag = "HRO",
            leader_id = 1,
            level = 5,
            emblem_color = new game.common.Color { r = 255, g = 215, b = 0, a = 255 },
            created_at = 1640000000
        };

        Assert(guild.name == "Heroes", "guild.name");
        Assert(guild.tag == "HRO", "guild.tag");

        var member = new game.social.GuildMember
        {
            guild_id = 1,
            player_id = 1,
            rank = game.social.GuildMember.Rank.Leader,
            joined_at = 1640000000
        };

        Assert(member.rank == game.social.GuildMember.Rank.Leader, "member.rank");

        Console.WriteLine("    PASS");
        passed++;
    }

    static void TestBinaryComplex()
    {
        Console.WriteLine("  Testing binary serialization of complex types...");

        var original = new game.character.Player
        {
            id = 999,
            name = "SerializationTest",
            level = 99,
            experience = 9999999,
            stats = new game.character.Stats
            {
                hp = 9999, max_hp = 9999,
                mp = 4999, max_mp = 4999,
                strength = 255, agility = 255, intelligence = 255, vitality = 255
            },
            position = new game.common.Vec3 { x = 123.456f, y = 789.012f, z = 345.678f },
            status = game.character.Player.Status.InBattle,
            guild_id = 42
        };

        // Serialize
        using var ms = new System.IO.MemoryStream();
        using var writer = new System.IO.BinaryWriter(ms);
        game.character.BinaryWriters.WritePlayer(writer, original);

        // Deserialize
        ms.Position = 0;
        using var reader = new System.IO.BinaryReader(ms);
        var loaded = game.character.BinaryReaders.ReadPlayer(reader);

        Assert(loaded.id == original.id, "id mismatch");
        Assert(loaded.name == original.name, "name mismatch");
        Assert(loaded.level == original.level, "level mismatch");
        Assert(loaded.stats.hp == original.stats.hp, "stats.hp mismatch");
        Assert(Math.Abs(loaded.position.x - original.position.x) < 0.001f, "position.x mismatch");
        Assert(loaded.status == original.status, "status mismatch");
        Assert(loaded.guild_id == original.guild_id, "guild_id mismatch");

        Console.WriteLine($"    PASS (serialized {ms.Length} bytes)");
        passed++;
    }

    static void Main()
    {
        Console.WriteLine("=== Test Case 08: Complex Schema ===");

        TestCommonTypes();
        TestCharacterTypes();
        TestItemTypes();
        TestSocialSystem();
        TestBinaryComplex();

        if (failed > 0)
        {
            Console.WriteLine($"=== {failed} tests failed! ===");
            Environment.Exit(1);
        }
        else
        {
            Console.WriteLine("=== All tests passed! ===");
        }
    }
}
